app/actions.py

```diff
@@ -92,30 +92,158 @@ def performDeleteRange(self, upperRow, upperCol, lowerRow, lowerCol):
+    penRow, penCol = bookmarkData['pen']
+    markerRow, markerCol = bookmarkData['marker']
+    selectionMode = bookmarkData['selectionMode']
+    self.cursorMoveAndMark(penRow - self.penRow, penCol - self.penCol,
+        markerRow - self.markerRow, markerCol - self.markerCol,
+        selectionMode - self.selectionMode)
+    self.redo()
 
   def bookmarkNext(self):
+    """
+    Goes to the closest bookmark after the cursor.
+
+    Args:
+      None.
+
+    Returns:
+      None.
+    """
     app.log.debug()
-    self.view.bookmarkIndex += 1
-    self.bookmarkGoto()
+    if not len(self.bookmarks):
+      self.setMessage("No bookmarks to jump to")
+      return
+    _, _, lowerRow, _ = self.startAndEnd()
+    tempBookmark = ((lowerRow, float('inf')),)
+    index = bisect.bisect(self.bookmarks, tempBookmark)
+    self.bookmarkGoto(self.bookmarks[index % len(self.bookmarks)])
 
   def bookmarkPrior(self):
+    """
+    Goes to the closest bookmark before the cursor.
+
+    Args:
+      None.
+
+    Returns:
+      None.
+    """
     app.log.debug()
-    self.view.bookmarkIndex -= 1
-    self.bookmarkGoto()
+    if not len(self.bookmarks):
+      self.setMessage("No bookmarks to jump to")
+      return
+    upperRow, _, _, _ = self.startAndEnd()
+    tempBookmark = ((upperRow,),)
+    index = bisect.bisect_left(self.bookmarks, tempBookmark)
+    bookmark = self.bookmarkGoto(self.bookmarks[index - 1])
 
   def bookmarkRemove(self):
+    """
+    Removes bookmarks in all selected lines.
+
+    Args:
+      None.
+
+    Returns:
+      None.
+    """
+    def helpRemoveBookmarks(bookmarkRange, startIndex, endIndex):
+      if endIndex < startIndex:
+        return
+      if startIndex == endIndex:
+        currentBookmarkRange = self.bookmarks[startIndex][0]
+        if self.bookmarksOverlap(bookmarkRange, currentBookmarkRange):
+          self.bookmarks[startIndex] = None
+        return
+      currentIndex = (startIndex + endIndex) / 2
+      prevBookmarkRange = self.bookmarks[currentIndex][0]
+      nextBookmarkRange = self.bookmarks[currentIndex + 1][0]
+      if self.bookmarksOverlap(bookmarkRange, prevBookmarkRange):
+        self.bookmarks[currentIndex] = None
+      if bookmarkRange[0] < prevBookmarkRange[0]: 
+        helpRemoveBookmarks(bookmarkRange, startIndex, currentIndex - 1)
+      if self.bookmarksOverlap(bookmarkRange, nextBookmarkRange):
+        self.bookmarks[currentIndex + 1] = None
+      if bookmarkRange[-1] > nextBookmarkRange[-1]:
+        helpRemoveBookmarks(bookmarkRange, currentIndex + 2, endIndex)
+
     app.log.debug()
-    return app.bookmarks.remove(self.view.bookmarkIndex)
+    upperRow, _, lowerRow, _ = self.startAndEnd()
+    numBookmarks = len(self.bookmarks)
+    bookmarkRange = (upperRow, lowerRow)
+    if self.bookmarks:
+      helpRemoveBookmarks(bookmarkRange, 0, numBookmarks - 1)
+      self.bookmarks = [b for b in self.bookmarks if b]
+    return len(self.bookmarks) != numBookmarks
```

Comment:
What do you think of changing this from a recursive sub-devision into a lookup and a linear walk?

Here's a hackish example
```
a = [(0, 5), (8, 10), (12, 12), (15, 20)]
tests = [(0, 0), (0, 10), (7, 10), (8, 20), (30, 30), (0, 500)]
print 'test list', a

def removeOverlaps(rangeList, needle):
  index = bisect.bisect_left(a, needle)
  if index and needle[0] <= rangeList[index-1][1]:
    index -= 1
  begin = end = index
  while True:
    index += 1
    end = index
    if index >= len(rangeList) or needle[1] < rangeList[index][0]:
      break
  return rangeList[:begin] + rangeList[end:]

for r in tests:
  print 'range', r, 'after remove', removeOverlaps(a, r)
```